### ISSUE TITLE
New field in monitors to auto include ready status

### ISSUE THREAD

**Issue by snuggie12** [prometheus-operator/prometheus-operator#4877]
**What is missing?**
Curious if there's any interest in adding a field to `PodMonitor` and `ServiceMonitor` that would automatically add relabeling which checks if a resource is ready. 

**Why do we need it?**
Some of our devs get scraped early and that is messing with our progressive delivery system. Adding relabelling that uses `__meta_kubernetes_endpoint_ready` can fix that issue. Not sure we _need_ it since we can get most of our devs to update easily enough, but figured I'd bring it up to see if others have interest in making it a built in feature.

**Environment**

* Prometheus Operator version:

    `prometheus-operator:v0.55.1`
    
    

**Anything else we need to know?**:

**Comment by snuggie12** [comment]
Yeah, it's similar but the endpoint maps back to servicemonitor which is what we use.

I also see loads of people have issues of targets not leaving where as this is about starting scraping too early. Didn't know if my ticket should be separate or not but happy to close and watch that issue if that'll track both podmonitor and servicemonitor.

**Comment by slashpai** [comment]
I am ok with keeping this issue open but we could follow same approach for both.
Would you like to contribute for the fix?

**Comment by simonpasquier** [comment]
I think that #4816 would also apply in your case but we can keep this one open until we have a clearer picture.

**Comment by adabuleanu** [comment]
I ended up on this thread because I have experienced a similar issue with a slow starting pod  ~ 2 minutes (with all the probes configured), so I ended up with a lot of 503 error codes until the pod transitioned into ready state. I expected that the default behavior should have been to scrape ready endpoints. Analysing this thread and  https://github.com/prometheus-operator/prometheus-operator/issues/4816, I think they do not achieve the same result. In https://github.com/prometheus-operator/prometheus-operator/issues/4816, if I set `__meta_kubernetes_pod_phase` to `Running` (according to the [Pod Lifecycle](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-phase) this phase include containers that are in the process of starting) , I will still experience my issue (my pod takes 2 minutes to start and if I have scraping configured for 30 seconds, I will get 3-4 503 error codes). But, if I configure `__meta_kubernetes_endpoint_ready`, the scraping will not start until the pod is [Ready](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-conditions) (which means it waits for all k8s condition + readinessProbe + startupProbe).

I have a question on configuring `__meta_kubernetes_pod_ready` / `__meta_kubernetes_pod_phase` / `__meta_kubernetes_endpoint_ready`: are there any concerns related to loosing metrics? Like for example in the case of OOM. Are there any side-effects? Surprisingly I did not find any docs on this, although I think a lot of people have this issue.

**Comment by JoaoBraveCoding** [comment]
@simonpasquier I think we might be able to close this one

**Comment by simonpasquier** [comment]
Right, closing it!
